src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java

```diff
@@ -48,81 +53,135 @@
+            String datacenter = getDatacenterSource.apply(peer);
+            peerToDatacenter.put(peer, datacenter);
+            datacenterToPeers.put(datacenter, peer);
+        }
+
+        // In the case where we do not want to block startup on remote datacenters (e.g. because clients only use
+        // LOCAL_X consistency levels), we remove all other datacenter hosts from the mapping and we only wait
+        // on the remaining local datacenter.
+        if (!blockForRemoteDcs)
+        {
+            datacenterToPeers.keySet().retainAll(Collections.singleton(localDc));
+            logger.info("Blocking coordination until only a single peer is DOWN in the local datacenter, timeout={}s",
+                        TimeUnit.NANOSECONDS.toSeconds(timeoutNanos));
+        }
+        else
+        {
+            logger.info("Blocking coordination until only a single peer is DOWN in each datacenter, timeout={}s",
+                        TimeUnit.NANOSECONDS.toSeconds(timeoutNanos));
+        }
 
         AckMap acks = new AckMap(3);
-        int target = (int) ((targetPercent / 100.0) * peers.size());
-        CountDownLatch latch = new CountDownLatch(target);
+        Map<String, CountDownLatch> dcToRemainingPeers = new HashMap<>(datacenterToPeers.size());
+        for (String datacenter: datacenterToPeers.keys())
+        {
+            dcToRemainingPeers.put(datacenter,
+                                   new CountDownLatch(Math.max(datacenterToPeers.get(datacenter).size() - 1, 0)));
+        }
+
+        long startNanos = System.nanoTime();
 
         // set up a listener to react to new nodes becoming alive (in gossip), and account for all the nodes that are already alive
-        Set<InetAddressAndPort> alivePeers = Sets.newSetFromMap(new ConcurrentHashMap<>());
-        AliveListener listener = new AliveListener(alivePeers, latch, acks);
+        Set<InetAddressAndPort> alivePeers = Collections.newSetFromMap(new ConcurrentHashMap<>());
+        AliveListener listener = new AliveListener(alivePeers, dcToRemainingPeers, acks, peerToDatacenter::get);
         Gossiper.instance.register(listener);
 
-        // send out a ping message to open up the non-gossip connections
-        sendPingMessages(peers, latch, acks);
+        // send out a ping message to open up the non-gossip connections to all peers. Note that this sends the
+        // ping messages to _all_ peers, not just the ones we block for in dcToRemainingPeers.
+        sendPingMessages(peers, dcToRemainingPeers, acks, peerToDatacenter::get);
 
         for (InetAddressAndPort peer : peers)
+        {
             if (Gossiper.instance.isAlive(peer) && alivePeers.add(peer) && acks.incrementAndCheck(peer))
-                latch.countDown();
+            {
+                String datacenter = peerToDatacenter.get(peer);
+                // We have to check because we might only have the local DC in the map
+                if (dcToRemainingPeers.containsKey(datacenter))
```

Comment:
I know some people who are totally going to be triggered by this since it does the lookup twice, but it is 100% fine. I think it's clearer this way. Get with null check is three lines instead of two. Really I just mention it so later if they come back and complain I can claim it was intentional :-)